graylog2-server/src/main/java/org/graylog2/entitygroups/rest/EntityGroupResource.java

```diff
@@ -0,0 +1,207 @@
+    @RequiresPermissions(ENTITY_GROUP_CREATE)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_CREATE)
+    public Response create(@ApiParam(name = "JSON Body") EntityGroupRequest request) {
+        try {
+            return Response.ok().entity(entityGroupService.create(request.toEntityGroup())).build();
+        } catch (DuplicateKeyException | MongoWriteException e) {
+            throw new BadRequestException(StringUtils.f("Entity group '%s' already exists", request.name()));
+        }
+    }
+
+    @PUT
+    @Path("/{id}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation("Update an entity group")
+    @ApiResponses({
+            @ApiResponse(code = 400, message = "An entity group already exists with id or name"),
+            @ApiResponse(code = 404, message = "No such entity group")
+    })
+    @RequiresPermissions(ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_UPDATE)
+    public Response update(@ApiParam(name = "id", required = true) @PathParam("id") String id,
+                           @ApiParam(name = "JSON Body") EntityGroupRequest request) {
+        if (!isPermitted(ENTITY_GROUP_EDIT, id)) {
+            throw new ForbiddenException("Not allowed to edit group id " + id);
+        }
+        try {
+            return Response.ok().entity(entityGroupService.update(id, request.toEntityGroup())).build();
+        } catch (IllegalArgumentException e) {
+            throw new NotFoundException(getRootCauseMessage(e), e);
+        } catch (DuplicateKeyException | MongoWriteException e) {
+            throw new BadRequestException(StringUtils.f("Entity group '%s' already exists", request.name()));
+        }
+    }
+
+    @PUT
+    @Path("/{group_id}/add_entity")
+    @ApiOperation("Add an entity to an entity group")
+    @ApiResponses(@ApiResponse(code = 404, message = "No such entity group"))
+    @RequiresPermissions(ENTITY_GROUP_EDIT)
+    @AuditEvent(type = AuditEventTypes.ENTITY_GROUP_UPDATE)
+    public Response addEntityToGroup(@ApiParam(name = "group_id", required = true) @PathParam("group_id") String groupId,
+                                     @ApiParam(name = "entity_type", required = true) @QueryParam("entity_type") String entityType,
+                                     @ApiParam(name = "entity_id", required = true) @QueryParam("entity_id") String entityId) {
+        if (!isPermitted(ENTITY_GROUP_EDIT, groupId)) {
+            throw new ForbiddenException("Not allowed to edit group id " + groupId);
+        }
+
+        try {
+            return Response.ok().entity(entityGroupService.addEntityToGroup(groupId, entityType, entityId)).build();
+        } catch (IllegalArgumentException e) {
+            throw new NotFoundException(getRootCauseMessage(e), e);
+        }
+    }
+
+    @PUT
+    @Path("/get_for_entities")
+    @ApiOperation("Get a list of entity groups for a list of entities")
+    @RequiresPermissions(ENTITY_GROUP_READ)
```

Comment:
same as with findPaginated, this needs to check the returned entities individually